### PR TITLE
PCHR-3846: Add leave report filter

### PR DESCRIPTION
**Pull request by reneolivo**
## Overview

This PR adds date filters to the admin's leave report.

## Before
leave reports hr17 9000 20

## After
anim
* Note: please ignore the date picker's border. See the comments for more information.

## Technical details

The way default values for the custom filters are generated changed since for the leave report they depend on the current period provided by the API. The new sequence is as follows:

* When loading the data for the report, also load the default values for the custom filter, if a `resolveCustomFilterDefaultValues` function is provided in the report configuration. This function must return a promise which resolves to an object containing the default values for each one of the inputs.
```js
// from the leave report configuration:
'resolveCustomFilterDefaultValues': function () {
  var today = moment().format(this.DEFAULT_DATE_FORMAT);

  return CRM.api3('AbsencePeriod', 'get', {
    'sequential': 1,
    'start_date': { '<=': today },
    'end_date': { '>=': today },
    'options': { 'limit': 1 }
  })
    .then(function (periods) {
      var period = _.first(periods.values);

      if (!period) {
        return;
      }

      // *from* and *to* are the name of the form inputs for the custom filters
      // in the leave report:
      return {
        from: period.start_date,
        to: period.end_date
      };
    });
},
```
* As explained before, this `resolveCustomFilterDefaultValues` is executed when loading the report data:
```js
// From the pivot table js file:
PivotTable.prototype.initPivotDataLoading = function () {
  // ...
  $.when(
    CRM.api3(apiCalls),
    this.resolveCustomFilterDefaultValues()
  )
    .then(function (results) {
      // ...
    });
};

PivotTable.prototype.resolveCustomFilterDefaultValues = function () {
  var dateInputsDefaultValues = this.getDefaultValuesForDateInputs();

  return $.when()
    .then(function () {
      return this.config.resolveCustomFilterDefaultValues
        ? this.config.resolveCustomFilterDefaultValues()
        : {};
    }.bind(this))
    .then(function (defaultValues) {
      // it also passes the date input default values so it's not necessary to define a
      // resolver just for date inputs of which their default is today:
      this.customFilterDefaultValues = _.extend({}, dateInputsDefaultValues, defaultValues);
    }.bind(this));
};

PivotTable.prototype.getDefaultValuesForDateInputs = function () {
  var defaultValues = {};
  var today = moment().format(this.DEFAULT_DATE_FORMAT);

  this.customFilterForm.find('.crm-ui-datepicker').each(function () {
    var inputName = $(this).attr('name');

    defaultValues[inputName] = today;
  });

  return defaultValues;
};
```
* After loading the pivot table data and the custom field default values, and during the post render sequence, the default values are assigned to the custom field inputs:
```js
PivotTable.prototype.initCustomFilterDefaultValues = function () {
  this.customFilterForm.find('input, select, textarea')
    .each(function (index, element) {
      var input = $(element);
      var inputName = input.attr('name');
      var inputDefaultValue = this.customFilterDefaultValues[inputName];

      if (!inputDefaultValue) {
        return;
      }

      input.on('change', function () {
        var isValueEmpty = _.isEmpty(input.val());

        if (isValueEmpty) {
          input.val(inputDefaultValue).change();
        }
      });
      input.change();
    }.bind(this));
};
```
If the input changes and the value is empty, it will use the default value instead.
* Finally, for the actual filtering, the `customFilter` function was defined as follows:
```js
'customFilter': function (record) {
  var dates = {
    start: moment(this.customFilterValues.from),
    end: moment(this.customFilterValues.to)
  };
  var request = {
    start: moment(record['Absence Start Date']),
    end: moment(record['Absence End Date'])
  };

  return request.start.isSameOrAfter(dates.start) && request.end.isSameOrBefore(dates.end);
}
```

## Comments

* For more information about report's custom filtering, please read the following PR:
https://github.com/compucorp/uk.co.compucorp.civicrm.pivotreport/pull/116
* This PR is related to this other PR: https://github.com/compucorp/org.civicrm.shoreditch/pull/111 since the later fixes an issue with the date picker borders.

### CODE REVIEW

File: js/PivotReport.PivotTable.js [hunk 1]
```diff
@@ -276,30 +292,36 @@ CRM.PivotReport.PivotTable = (function ($) {
...
-      that.relativeFilters = result.relativeFilters.values;
-      that.header = result.getHeader.values;
-      that.total = parseInt(result.getCount.values, 10);
-      that.pivotCount = parseInt(result.getPivotCount.values, 10);
-      that.crmConfig = result.getConfig.values[0];
-
-      $.each(that.dateFields, function (i, value) {
-        that.config.derivedAttributes[value + ' (' + ts('per month') + ')'] = $.pivotUtilities.derivers.dateFormat(value, '%y-%m');
-      });
+    $.when(
```
Comment (deb1990) [inline js/PivotReport.PivotTable.js]: Is it necessary to resolve `this.resolveCustomFilterDefaultValues()`before calling the done callback? I mean is it dependant?

File: templates/CRM/PivotReport/Page/LeaveReport.tpl [hunk 2]
```diff
@@ -100,7 +116,41 @@
           return 'No';
         }
       },
-      'hiddenAttributes': ['Is TOIL', 'Absence Amount', 'Absence Calculation Unit', 'Is TOIL']
+      'hiddenAttributes': ['Is TOIL', 'Absence Amount', 'Absence Calculation Unit', 'Is TOIL'],
```
[the remark this answers is not in the document]
Comment (reneolivo) [reply templates/CRM/PivotReport/Page/LeaveReport.tpl]: good catch, it should be added once only.

File: js/PivotReport.PivotTable.js [hunk 3]
```diff
@@ -84,6 +84,33 @@ CRM.PivotReport.PivotTable = (function ($) {
...
 
+  /**
+   * Fetched some required data from the api including the records count, date
+   * fields, headers, etc.
+   *
+   * @return {Promise} resolves to an object containing each api call result.
+   */
+  PivotTable.prototype.fetchRequiredApiData = function () {
+    var countParams = this.config.getCountParams();
+    countParams.entity = this.config.entityName;
```
Comment (deb1990) [inline js/PivotReport.PivotTable.js]: Please place it after defining all variables.

**Comment by tunbola** [comment]
Not needed anymore.

**Pull request closed without merging**